### PR TITLE
[AArch64] Add soft-float ABI

### PR DESCRIPTION
**Pull request by ostannard**
This adds support for the AArch64 soft-float ABI. The specification for this ABI is currently in review at https://github.com/ARM-software/abi-aa/pull/232, and I won't commit this until that PR is merged.

Because all existing AArch64 hardware has floating-point hardware, we expect this to be a niche option, only used for embedded systems on R-profile systems. We are going to document that SysV-like systems should only ever use the base (hard-float) PCS variant: https://github.com/ARM-software/abi-aa/pull/233. For that reason, I've not added an option to select the ABI independently of the FPU hardware, instead the new ABI is enabled iff the target architecture does not have an FPU.

For testing, I have run this through an ABI fuzzer, but since this is the first implementation it can only test for internal consistency (callers and callees agree on the PCS), not for conformance to the ABI spec.

### CODE REVIEW

**Comment by DavidSpickett** [comment]
(maybe this is for a later patch but...)

Does this patch include the clang side options for the ABI or does `-march=...+no-fp-armv8` already work for that?

**Comment by ostannard** [comment]
This doesn't add any new options, and I'd like to avoid that if possible, instead it is turned on by `-march=...+nofp`.

**Comment by DavidSpickett** [comment]
That makes sense. Didn't realise we had an `fp` flag, I don't see it in `llvm/include/llvm/TargetParser/AArch64TargetParser.h`.

**Comment by efriedma-quic** [comment]
I'm a little surprised we don't need any LLVM backend changes, but I guess in soft-float mode we basically treat floats as ints anyway, so maybe things just work.  I'd like to see appropriate regression tests, though (if they don't already exist).

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 1]
```diff
@@ -534,7 +540,8 @@ Address AArch64ABIInfo::EmitAAPCSVAArg(Address VAListAddr, QualType Ty,
     BaseTy = ArrTy->getElementType();
     NumRegs = ArrTy->getNumElements();
   }
-  bool IsFPR = BaseTy->isFloatingPointTy() || BaseTy->isVectorTy();
+  bool IsFPR = Kind == AArch64ABIKind::AAPCS &&
```
Comment (davemgreen) [inline clang/lib/CodeGen/Targets/AArch64.cpp:543]: Could this be Kind != AArch64ABIKind::AAPCS` in case other ABIKinds make their way here.

**Comment by ostannard** [comment]
> I'm a little surprised we don't need any LLVM backend changes, but I guess in soft-float mode we basically treat floats as ints anyway, so maybe things just work. I'd like to see appropriate regression tests, though (if they don't already exist).

Yes, the backend already treated floats as if they were an integer of the same width, so no changes are needed there. I've added a test for the backend.

**Comment by ktkachov-arm** [comment]
Changing the ABI through the architecture flags i.e. +nofp has potential to greatly confuse users, who may come from other architectures that have soft-float ABIs already (including 32-bit arm)

- Does this also happen when using +nosimd?
- What about users like the Linux kernel that build with -mgeneral-regs-only?
- There is already an -mabi= option. How come it doesn't affect what is a pretty major ABI decision?

**Comment by ostannard** [comment]
I originally implemented it this way to avoid making two ABIs valid for one target, but I agree that it would be confusing to have the ABI change automatically. How about we use the `-mabi=` option as you suggest, with values `aapcs` (already accepted by clang) and `aapcs-soft` (new)?

* `-mabi=aapcs` is always the default
* If `-mabi=aapcs` (or the default) is used without an FPU, then floating-point argument and return values are rejected with an error. This matches GCC's current behaviour, and would be a change for clang, though clang previously used a non-compliant ABI. Code which is accepted with this combination is compatible with both ABIs.
* If `-mabi=aapcs-soft` is used with an FPU, then we reject the command-line options with an error. This avoids having two incompatible ABIs for one target. We will always have the option to relax this in future without breaking existing code, but I'd like to avoid that unless there's a very good reason.

> Does this also happen when using +nosimd?
No, in clang `+nosimd` leaves the FPU turned on. That's probably a bad thing, since that isn't a valid combination for AArch64, but that's a different issue.

> What about users like the Linux kernel that build with -mgeneral-regs-only?
In clang, that has the same behaviour as `+nofp`. I vaguely remember something about GCC still allowing FP instructions in inline assembly with this option, but we don't do that in clang because the assembler is integrated into the compiler, and gets the same target options. For the purposes of the ABI, I think we should treat it the same way as `+nofp`.

**Comment by ostannard** [comment]
Ping

**Comment by ostannard** [comment]
Ping

**Comment by ostannard** [comment]
Ping, the ABI specification change (https://github.com/ARM-software/abi-aa/pull/232) has now been merged.

**Review by stuij — APPROVED** [review]
This seems like a sensible and unobtrusive solution to me. Also the change is backed by a change in the Arm ABI, and it looks like the other review comments have been addressed. I've also built this change locally and ran the lit tests, and everything passes.

**Comment by ostannard** [comment]
One test caused a buildbot failure when the AArch64 backend is not built, fixed by 5b8e7ed787f6e537876c4fdafd070eba9681f343

**Comment by nathanchance** [comment]
For what it's worth, this breaks building the Linux kernel for `ARCH=arm64` pretty badly, with errors in several drivers:

```
$ make -skj"$(nrpoc)" ARCH=arm64 LLVM=1 mrproper defconfig all
...
drivers/clk/qcom/gcc-ipq6018.c:896:2: error: expression requires 'double' type support, but ABI 'aapcs' does not support it
  896 |         F(533000000, P_GPLL0, 1.5, 0, 0),
      |         ^
drivers/clk/qcom/clk-rcg.h:10:41: note: expanded from macro 'F'
   10 | #define F(f, s, h, m, n) { (f), (s), (2 * (h) - 1), (m), (n) }
      |                                         ^
...
In file included from drivers/gpu/drm/msm/adreno/a2xx_gpu.c:4:
In file included from drivers/gpu/drm/msm/adreno/a2xx_gpu.h:13:
drivers/gpu/drm/msm/adreno/a2xx.xml.h:1849:24: error: 'A2XX_PA_CL_VPORT_XSCALE' requires 'float' type support, but ABI 'aapcs' does not support it
 1849 | static inline uint32_t A2XX_PA_CL_VPORT_XSCALE(float val)
      |                        ^
drivers/gpu/drm/msm/adreno/a2xx.xml.h:1849:24: note: 'A2XX_PA_CL_VPORT_XSCALE' defined here
drivers/gpu/drm/msm/adreno/a2xx.xml.h:1857:24: error: 'A2XX_PA_CL_VPORT_XOFFSET' requires 'float' type support, but ABI 'aapcs' does not support it
 1857 | static inline uint32_t A2XX_PA_CL_VPORT_XOFFSET(float val)
      |                        ^
drivers/gpu/drm/msm/adreno/a2xx.xml.h:1857:24: note: 'A2XX_PA_CL_VPORT_XOFFSET' defined here
drivers/gpu/drm/msm/adreno/a2xx.xml.h:1865:24: error: 'A2XX_PA_CL_VPORT_YSCALE' requires 'float' type support, but ABI 'aapcs' does not support it
 1865 | static inline uint32_t A2XX_PA_CL_VPORT_YSCALE(float val)
      |                        ^
drivers/gpu/drm/msm/adreno/a2xx.xml.h:1865:24: note: 'A2XX_PA_CL_VPORT_YSCALE' defined here
...
```

I suspect this may be related to the kernel's use of `-mgeneral-regs-only`? Up until this point though, there have been no reported issues with code generation or runtime time impact with `clang`-built kernels.

**Comment by Prabhuk** [comment]
This commit breaks building Fuchsia's Kernel as well.

Sample builder log: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8755980051536646593/+/u/build/ninja/stdout

```
[87049/291900](68) CXX kernel.phys_arm64/obj/src/lib/elfldltl/libelfldltl.leb128.cc.o
FAILED: kernel.phys_arm64/obj/src/lib/elfldltl/libelfldltl.leb128.cc.o
../../prebuilt/third_party/clang/custom/bin/clang++ -MD -MF kernel.phys_arm64/obj/src/lib/elfldltl/libelfldltl.leb128.cc.o.d -o kernel.phys_arm64/obj/src/lib/elfldltl/libelfldltl.leb128.cc.o -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -DZX_ASSERT_LEVEL=2 -DLK_DEBUGLEVEL=2 -DWITH_FRAME_POINTERS=1 -I../../zircon/system/public -I../../src/lib/elfldltl/incl...
In file included from ../../src/lib/elfldltl/leb128.cc:5:
In file included from ../../src/lib/elfldltl/include/lib/elfldltl/dwarf/encoding.h:8:
In file included from ../../sdk/lib/stdcompat/include/lib/stdcompat/span.h:13:
In file included from ../../sdk/lib/stdcompat/include/lib/stdcompat/iterator.h:10:
In file included from ../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/iterator:683:
In file included from ../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__iterator/common_iterator.h:31:
In file included from ../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/variant:220:
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:437:9: error: '__v' requires 'float' type support, but ABI 'aapcs' does not support it
  437 |     if (__v == 0.0f)
      |         ^
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:435:49: note: '__v' defined here
  435 |   _LIBCPP_HIDE_FROM_ABI size_t operator()(float __v) const _NOEXCEPT {
      |                                                 ^
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:439:34: error: 'operator()' requires 'float' type support, but ABI 'aapcs' does not support it
  439 |     return __scalar_hash<float>::operator()(__v);
      |                                  ^
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:249:32: note: 'operator()' defined here
  249 |   _LIBCPP_HIDE_FROM_ABI size_t operator()(_Tp __v) const _NOEXCEPT {
      |                                ^
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:435:32: error: 'operator()' requires 'float' type support, but ABI 'aapcs' does not support it
  435 |   _LIBCPP_HIDE_FROM_ABI size_t operator()(float __v) const _NOEXCEPT {
      |                                ^
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:435:32: note: 'operator()' defined here
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:447:9: error: '__v' requires 'double' type support, but ABI 'aapcs' does not support it
  447 |     if (__v == 0.0)
      |         ^
../../prebuilt/third_party/clang/custom/bin/../include/c++/v1/__functional/hash.h:445:50: note: '__v' defined here
  445 |   _LIBCPP_HIDE_FROM_ABI size_t operator()(double __v) const _NOEXCEPT {
      |             ...
```

**Comment by Prabhuk** [comment]
@ostannard -- Can you please fix or submit a revert? Thank you.

**Pull request merged**